src/box/iproto.cc

```diff
@@ -2316,6 +2444,67 @@ iproto_thread_init(struct iproto_thread *iproto_thread)
 	return 0;
 }
 
+static void
+tx_send_shutdown(struct cmsg *m)
+{
+	struct iproto_connection *con =
+		container_of(m, struct iproto_connection, shutdown_msg);
+	struct session *session = con->session;
+	if (session == NULL || !session->graceful_shutdown) {
+		shutdown(con->input.fd, SHUT_RD);
+	} else {
+		iproto_reply_shutdown(con->tx.p_obuf, ::schema_version);
+		if (!con->tx.is_push_sent)
+			tx_begin_push(con);
+		else
+			con->tx.is_push_pending = true;
+	}
+}
+
+static void
+iproto_process_shutdown(struct iproto_thread* iproto_thread)
+{
+	while (! rlist_empty(&iproto_thread->to_shutdown_connections)) {
+		struct iproto_connection *con =
+			rlist_first_entry(&iproto_thread->to_shutdown_connections,
+					  struct iproto_connection,
+					  to_shutdown_list);
+		rlist_del(&con->to_shutdown_list);
+		rlist_add_tail_entry(&con->iproto_thread->to_close_connections,
+				     con, to_close_list);
+		cmsg_init(&con->shutdown_msg, iproto_thread->send_shutdown_route);
+		cpipe_push(&iproto_thread->tx_pipe, &con->shutdown_msg);
+	}
+}
+
+static void
+iproto_do_close_all_connections(struct iproto_thread *iproto_thread)
+{
+	while(! rlist_empty(&iproto_thread->to_close_connections)) {
+		struct iproto_connection *con =
+			rlist_first_entry(&iproto_thread->to_close_connections,
+					  struct iproto_connection,
+					  to_close_list);
+		rlist_del(&con->to_close_list);
+		iproto_connection_close(con);
+	}
+}
+
+static inline void
+iproto_send_stop_msg(void);
+
+static inline void
+iproto_send_shutdown_msg(void);
+
+static int
+graceful_shutdown(void *arg)
+{
+	(void) arg;
+	iproto_send_shutdown_msg();
+	iproto_send_stop_msg();
```

Comment:
You move connections from `to_shutdown_list` to `to_close_list` in `iproto_send_shutdown_msg`. If client
connect between `iproto_send_shutdown_msg` and `iproto_send_stop_msg` this connection not moved.
And server fail to graceful shutdown.
Server:
```lua
os.execute("rm -r *.snap *.xlog")

box.cfg {
        iproto_threads = 1,
        listen = "/tmp/1"
}
box.schema.user.grant('guest','execute,write,read','universe')
box.ctl.set_on_shutdown_timeout(20)
require('console').start()
```
Client:
```lua
fiber = require('fiber')
os.execute("rm -rf *.snap *.xlog")
ccc = 10000
net_box = require('net.box')
table = {}
for i = 1, ccc do
    graceful_con = net_box.connect("/tmp/1")
    -- Next string is optional.
    graceful_con.space._session_settings:update('graceful_shutdown$
    table[i] = graceful_con
end
require('console').start()
```
If you run server, then client and then terminate server Ctrl + C, before client finish it's loop. Server often
fails with `on_shutdown triggers are timed out: not all triggers might have finished yet`, because
`builtin.iproto_wait_close_all_connections()` never finished.

Reply:
Added to the test